EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -545,3 +295,329 @@ void disconnectContactors(uint8_t isContactorError)
+        toAlsoBeNamed->prevState = toAlsoBeNamed->state;
+        toAlsoBeNamed->state = DISCHARGE_ENABLE_CHECK;
+    }
+
+
+    // If charge contactor should be on, but isn't, disconnect
+    // or if discharge contactor should be on, but isn't, disconnect
+    if (chargeContactorShouldBeOnButNot(chargeSense, isChargeTurningOn) || dischargeContactorShouldBeOnButNot(dischargeSense, isDischargeTurningOn))
+    {
+        disconnectContactors(1);
+        toAlsoBeNamed->state = CONTACTOR_DISCONNECTED;
+    }
+}
+
+void blocked(ToBeNamed* toAlsoBeNamed)
+{
+    const uint8_t commonSense = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+    const uint8_t chargeSense = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+    const uint8_t dischargeSense = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+
+    // If orion isn't overiding the contactors, but they are supposed to have been set, but aren't
+    // , this is an error condition and all contactors should be disconnected.
+    if (toAlsoBeNamed->prevState > COMMON_ENABLE_CHECK && 
+        (!commonSense || isChargeFalsePositive(chargeSense, toAlsoBeNamed->hasChargeBeenSet) || 
+        isDischargeFalsePositive(dischargeSense, toAlsoBeNamed->hasDischargeBeenSet)))
+    {
+        disconnectContactors(1);
+        toAlsoBeNamed->state = CONTACTOR_DISCONNECTED;
+    }
+    else if (orionStatus.batteryVoltagesInRange)
+    {
+        if (orionStatus.chargeContactorOverriden)
+        {
+            toAlsoBeNamed->hasChargeBeenSet = 0;
+        }
+
+        if (orionStatus.dischargeContactorOverriden)
+        {
+            toAlsoBeNamed->hasDischargeBeenSet = 0;
+        }
+
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.chargeContactorOverriden = 0;
+        orionStatus.dischargeContactorOverriden = 0;
+
+        osMutexRelease(orionStatus.orionStatusMutex);
+        toAlsoBeNamed->state = FIRST_CHECK;
+    }
+}
+
+const uint8_t isDischargeFalsePositive(const uint8_t dischargeSense, uint8_t hasDischargeBeenSet)
+{
+    return (!orionStatus.dischargeContactorOverriden && hasDischargeBeenSet && !dischargeSense);
+}
+
+const uint8_t isChargeFalsePositive(const uint8_t chargeSense, uint8_t hasChargeBeenSet)
+{
+    return (!orionStatus.chargeContactorOverriden && hasChargeBeenSet && !chargeSense);
+}
+
+const uint8_t canTurnOnDishargeContactor(const uint8_t dischargeSense, uint8_t hasDischargeBeenSet)
+{
+    return (orionStatus.allowDischarge && orionStatus.dischargeContactorOverriden && !dischargeSense);
+}
+
+const uint8_t canTurnOnChargeContactor(const uint8_t chargeSense)
+{
+    return (orionStatus.allowCharge && orionStatus.chargeContactorOverriden && !chargeSense);
+}
+
+const uint8_t dischargeContactorShouldBeOnButNot(const uint8_t dischargeSense, const uint8_t isDischargeTurningOn)
```

Comment:
Now that I think of it, both this function and `isDischargeFalsePositive` both check if the contactor disconnected, this one just checks it during the DONE state and the other one checks if the contactor disconnected during the CHECK/BLOCK phases. The names should probably reflect it, in a way that makes sense to someone that might not necessarily know the states in the code.
Maybe `dischargeDisconnectedDuringContactorSetting` for `falsePositive` and `dischageDisconnectedUnexpectedly` for `shouldBeOnButNot`? still sounds a little janky haha

Reply:
I think the names you've suggested make sense. They're janky, but at least there's no ambiguity haha